lib/NUPACK/emscripten/FullEnsemble.cpp

```diff
@@ -331,10 +368,160 @@ std::string getDotParens(bool pseudoknotted, const int seqlength, oneDnaStruct *
     {
        return singlestructure;
     }
-    
+}
 
+void getEnsembleDefect(char* seqChar, char* dotParensStructure, int temperature, bool pseudoknot, bool multiFold, double *returnEnsembleDefect, bool mfeDefect)
+{
+  //set global constants needed for correct folding and ensemble reporting     
     
+  DANGLETYPE=1;  
+
+  if (multiFold==true)
+  {
+    Multistranded = 1;
+  }
+  else
+  {
+    Multistranded = 0;
+  }
+
+  if (pseudoknot == true)
+  {
+    DO_PSEUDOKNOTS = 1;
+  }
+  else
+  {
+    DO_PSEUDOKNOTS = 0;
+  }
+
+  if (mfeDefect==true)
+  {
+    USE_MFE = 1;
+    // -degenerate flag not used for defect calcs, force ONLY_ONE_MFE
+    ONLY_ONE_MFE = 1;
+  }
+  else
+  {
+    USE_MFE = 0;
+    ONLY_ONE_MFE = 0;
+  }
+
+  int i, j;
+  int trySymetry=TRUE;
+  
+  
+  DBL_TYPE nsStar;
+  DBL_TYPE mfe; // Minimum free energy (not really used)
+  DBL_TYPE ene; // Free energy (used to check if the structure is legal)
+  int vs;
+  int complexity;
+  int tmpLength;
+  int seqlength;
+  int nbases; // Number of bases as read from ppairs or mfe file
+ 
+  int nNicks;  // Number of strands
+  int doCalc; // Whether we need to compute the pair probability matrix/mfe or not
+  
+   int seqStringLength;
+  char *tok; // Token
+  char tokseps[] = " \t\n"; // Token separators
+
+  //convert structure and sequence to char
 
+  
+  int seqNum[MAXSEQLENGTH+1];
+  seqlength = seqStringLength = tmpLength = strlen(seqChar);
+  
+
+  convertSeq(seqChar, seqNum, seqStringLength);
+
+  
+  // Get the number of strand breaks
+  nNicks = 0;
+  for (i = 0; i < tmpLength; i++) {
+    if (seqChar[i] == '+') {
+      nNicks++;
+      seqlength--;
+    }
+  }
+
+  int thepairs[MAXSEQLENGTH+1];
+  getStructureFromParens( dotParensStructure, thepairs, seqlength);
+  
+  ene = naEnergyPairsOrParensFullWithSym( thepairs, NULL, seqNum, RNA, 1 /*DANGLETYPE*/,
+          temperature, trySymetry,
+          SODIUM_CONC, MAGNESIUM_CONC,
+          USE_LONG_HELIX_FOR_SALT_CORRECTION);
+
+  // Allocate memory for storing pair probabilities
+   pairPr = (DBL_TYPE*) calloc( (seqlength+1)*(seqlength+1), sizeof(DBL_TYPE));
+  
+  //stuff for folding and getting probs
+    dnaStructures mfeStructs = {NULL, 0, 0, 0, NAD_INFINITY};
+  
+ //printInputs( argc, argv, seqChar, vs, NULL, parens, "screen");
+
+    if (USE_MFE) {
```

Comment:
The logic here was derived from `src/thermo/basics/defect.c`, right? How about adding a comment to reference that's where it came from?